### PR TITLE
simulate_molecule

### PR DESCRIPTION
**Pull request by Gabe-DiAntonio**
I wrote the functions we talked about yesterday. 
One question: how do I test a function that involves a random process?

### CODE REVIEW

File: src/Diffusion.jl [hunk 1]
```diff
@@ -1,9 +1,110 @@
...
+function absorb_boundary(pos,r) #models an absorbing boundary-pos inside the boundary stay, but pos outside disappear 
+    if(sum(abs2, pos) < r^2)
+        return pos
+    else
+        return nothing
+    end
+end
+
+function topolar(pos) #turns a cartesian coordinate into a polar coordinate 
+    if pos[1] >= 0 && pos[2] >= 0
```
Comment (timholy) [inline src/Diffusion.jl:23]: Julia has `theta = atan(y, x)` so that you don't have to check all these cases. But nice work, and if you prefer it's fine to leave this way.

File: src/Diffusion.jl [hunk 2]
```diff
@@ -1,9 +1,110 @@
...
+        theta = atan(pos[2]/pos[1])
+    elseif pos[1] >=0 && pos[2] <= 0
+        theta = atan(pos[2]/pos[1]) + 2π
+    else
+        theta = atan(pos[2]/pos[1]) + π
+    end
+    r = sqrt(sum(abs2, pos))
+    return SA[r,theta]
```
Comment (timholy) [inline src/Diffusion.jl:31]: One risk is that `[r, theta]` is "just a vector" and consequently could be interpreted as cartesian coordinates (it's not obvious by inspection which one it is, you just have to know...). If you wanted to clarify the meaning, you could define a `struct Polar` (hint: there's already one in [CoordinateTransformations](https://github.com/JuliaGeometry/CoordinateTransformations.jl)).

But I'm fine with leaving this as-is if you prefer it this way, I just thought I'd mention it.

File: src/Diffusion.jl [hunk 3]
```diff
@@ -1,9 +1,110 @@
...
+            return molhistory
+        else
+        push!(molhistory, pos)
+        end
+    end 
+    return molhistory
+end
+
+function nmolreflect(σ, steps, pos, r, molecules) #simulates multiple molecules at once-reflective
+    molhistories = []
```
Comment (timholy) [inline src/Diffusion.jl:95]:
```suggestion
    molhistories = Vector{typeof(pos)}[]
```
That should improve performance of code that analyzes these, you're essentially letting Julia's compiler know in advance what the types will be and therefore it will generate optimized code. Of my comments, this is the one I'd rate as better to change now before merging (the others are optional).

This will change your `Vector{Any}` test result, of course.

**Review by timholy — COMMENTED** [review]
This looks quite good, just a few minor comments below.

With respect to testing code with randomness, this is often difficult. What I typically do is test statistical properties, but this gets you towards some of the analysis code. (So you can do it in a separate pull request.) For example, you expect the following:
- with many molecules released from the same point, the variance should initially increase as `2Dt`
- with many molecules released near the edge with reflecting boundary conditions, over time the average position should be near the center of the disc
- with absorbing boundary conditions, the closer your initial release point is to the edge, the faster they will disappear.

But again, we can wait until you've written a bit of analysis code before adding those tests.

File: src/Diffusion.jl (hunk 1 above)
Comment (Gabe-DiAntonio) [reply src/Diffusion.jl:23]: Is the code slower if I leave it this way?

Comment (Gabe-DiAntonio) [reply src/Diffusion.jl:23]: Also-how do I make Julia do it for me? When I switched the code to theta = atan(y, x) it started to fail a bunch of tests.

Comment (timholy) [reply src/Diffusion.jl:23]: I don't see those failures: `theta = atan(pos[2], pos[1])` also passes tests. But let's leave it this way now.
